src/shogun/clustering/KMeans.cpp

```diff
@@ -77,35 +77,31 @@ void CKMeans::set_initial_centers(SGMatrix<float64_t> centers)
 
 void CKMeans::set_random_centers(SGVector<float64_t> weights_set, SGVector<int32_t> ClList, int32_t XSize)
 {
+	/*initialise the cluster centroids randomly among all data*/
 	CDenseFeatures<float64_t>* lhs=
 		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
 
-	for (int32_t i=0; i<XSize; i++)
+	SGVector<int32_t> Perm=SGVector<int32_t>(XSize);
+	SGVector<int32_t>::range_fill_vector(Perm, XSize, 0);
+	CMath::permute(Perm);
+
+	for (int32_t i=0; i<k; i++)
 	{
-		const int32_t Cl=CMath::random(0, k-1);
+		const int32_t Cl=Perm[i];
 		weights_set[Cl]+=1.0;
-		ClList[i]=Cl;
+		ClList[Cl]=Cl;
 
 		int32_t vlen=0;
 		bool vfree=false;
-		float64_t* vec=lhs->get_feature_vector(i, vlen, vfree);
+		float64_t* vec=lhs->get_feature_vector(Cl, vlen, vfree);
 
 		for (int32_t j=0; j<dimensions; j++)
-			mus.matrix[Cl*dimensions+j] += vec[j];
+			mus.matrix[i*dimensions+j] += vec[j];
```

Comment:
SGMatrix has the () operator which you can give row and column, could you use this here? makes the code easier to read. You see you have to clean up other peoples mess ;)